lib/trino-memory-context/src/main/java/io/trino/memory/context/MemoryAllocationValidator.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.memory.context;
+
+public interface MemoryAllocationValidator
```

Comment:
This is really similar to `MemoryReservationHandler`. Could that interface do the job? It seems that memory allocations via `reserveMemory` should return `Future` as it's caller intention is to wait for memory. The other method `tryReserveMemory` is for trying to allocate memory.

With interface like this, I don't see a reason for two methods to exist (other than exception customization).

Reply:
`Future<>` as a return type does not suite me well. No use for that. Sole purpose of validator is to "reject" the memory allocation if limit is exceeded. It is not handling the case when the limit is not exceeded but node run out of memory (pool is depleted).

Reply:
> Future<> as a return type does not suite me well. No use for that. Sole purpose of validator is to "reject" the memory allocation if limit is exceeded. It is not handling the case when the limit is not exceeded but node run out of memory (pool is depleted).

Ok, so why do we need an interface with two methods?

> Two methods there so we do not need to do exception driven logic in ChildAggregatedMemoryContext.tryUpdateBytes where MemoryAllocationValidator.tryReserveMemory is used.
If I just had reserveMemory I would need to try/catch there. It is an option, but I think it is not ellegant.

It seems that we leak implementation details to interface. I don't think two methods are needed now that you have `ValidatingLocalMemoryContext`

Reply:
I can make it one method which would return `Optional<String>`
* empty if allocation is valid
* containing allocation error description if allocation violates constraints
Then I can do the throwing in `ValidatingLocalMemoryContext`

Reply:
Actually, I take that back. Then the `ValidatingLocalMemoryContext` need to decide on exception type to throw. I am not a big fan of hardcoding the `ExceededMemoryLimitException(EXCEEDED_LOCAL_MEMORY_LIMIT,` there. It makes class less reusable. And by design nothing is enforcing us to use the class only for TaskContext.

I suggest we keep interface as proposed.